File: session/01_languages/erlang/destructuring.erl
#!/usr/bin/env escript

-record (name, {first, last, salutation}).

main([]) ->
  % if a variable doesn't have a value already, 
  % if it isn't bound, then equality gives it a value.
  A = 10,
  io:format("~p~n", [A]),
  % But if it is already bound,  then equality is a 
  % check on its value.  it's a match to see:  
  % is this thing  on the left equal to this thing on the right?
  % A = 20,
  
  % Destructuring List
  List = [1,2,3,4,5,6],
  [First,Second|Rest] = List,
  io:format("~p, ~p, ~p~n", [First, Second, Rest]),

  % Destructuring n-Tuples
  {X, Y} = {10, 20}, 
  io:format("~p, ~p~n", [X, Y]),

  % Destructuring Map (skip this)
  Map = #{ "a" => 2, "b" => 4, "c" => 8 },
  #{ "a" := V1, "c" := V3 } = Map,
  io:format("~p, ~p~n", [V1, V3]),

  % Destructuring records
  Name = #name { first = "Dhaval", last = "Dalal", salutation = "Mr."},
  io:format("~p~n", [Name]),
  #name{ first = FName, last = LName} = Name,
  io:format("~p ~p~n", [FName, LName]).  

File: session/01_languages/javascript/destructuring.js
// In FP, Destructuring is about extracting data from the innards of any structure.
// Destructuring list (or use as n-Tuple)
// Show this first
const list = [1,2,3,4,5,6]
let [first, second] = list
console.info("first = ", first);
console.info("second = ", second);
// Destructuring gives us a short-hand way of naming parts of the data-structure.
// then show by Adding to the above ...rest
[first, second, ...rest] = list
console.info("rest = ", rest);

// Destructuring Object
const name = {first: "Dhaval", last: "Dalal", salutation: 'Mr.'};
const {first: firstName, last: lastName} = name;
console.info(firstName);
console.info(lastName);

// Destructuring Function params
function capitalize({first: firstName, last: lastName}){
  return firstName.toUpperCase() + " " + lastName.toUpperCase();
}
console.info(capitalize(name));

// Destructuring Function params
function midpoint([x1,y1], [x2,y2] = [0,0]) { 
  function mid(a1, a2) { return (a1 + a2) / 2; }
  return [mid(x1, x2), mid(y1, y2)];
}

console.info(midpoint([2,3], [4,5]));
console.info(midpoint([2,3]));

// No need to show this - skip.
// Destructuring Map - No direct way
const map = new Map()
  .set('a', 2)
  .set('b', 4)
  .set('c', 8);
  
console.info(map);
// so you spread the map to check what is represented as internally
console.info(...map);
// it's list, so we destructure the list.
const [[fKey, fVal], [sKey,sVal], [tKey,tVal]] = map
console.info(fKey, fVal, sKey, sVal, tKey, tVal);
